src/main/java/com/limechain/storage/block/tree/Node.java

```diff
@@ -0,0 +1,210 @@
+package com.limechain.storage.block.tree;
+
+import io.emeraldpay.polkaj.types.Hash256;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Node represents element in the block tree
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Node {
+    private byte[] hash; // Block hash
+    private Node parent; // Parent node
+    private List<Node> children = new ArrayList<>(); // Nodes of children blocks
+    private long number; // block number
+    private Instant arrivalTime; // Time of arrival of the block
+    private boolean isPrimary; // whether the block was authored in a primary slot or not
+
+    public Node(byte[] hash, Node parent, long number) {
+        this.hash = hash;
+        this.parent = parent;
+        this.number = number;
+        arrivalTime = Instant.now();
+    }
+
+    /**
+     * Adds a child to the node
+     * @param node child to be added node
+     */
+    public void addChild(final Node node) {
+        this.children.add(node);
+    }
+
+    /**
+     * @return stringified hash and number of node
+     */
+    @Override
+    public String toString() {
+        return String.format("{hash: %s, number: %d, arrivalTime: %s}", new Hash256(hash), number, arrivalTime);
+    }
+
+    /**
+     * Recursively searches for a node with a given hash
+     * @param hash hash of the node
+     * @return node with the given hash
+     */
+    public Node getNode(final byte[] hash) {
+        if (Arrays.equals(this.hash, hash)) {
+            return this;
+        }
+        for (Node child : this.children) {
+            Node node = child.getNode(hash);
+            if (node != null) {
+                return node;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Recursively searches for all nodes with a given number
+     * @param number number of the block
+     * @return list of nodes with the given number
+     */
+    public List<byte[]> getNodesWithNumber(final long number) {
+        List<byte[]> hashes = new ArrayList<>();
+        for (Node child : children) {
+            if (child.number == number) {
+                hashes.add(child.hash);
+            }
+            if (child.number > number) {
+                return hashes;
+            }
+            hashes.addAll(child.getNodesWithNumber(number));
+        }
+        return hashes;
+    }
+
+    /**
+     * Traverse the tree following the parent nodes to verify if the given node is a descendant of the parent
+     * @param parent parent node to be traversed
+     * @return true if the node is a descendant of the parent
+     */
+    public boolean isDescendantOf(final Node parent) {
```

Comment:
I think that this method can be extracted into an interface. I think I saw it being used in another class as well. This is also applicable to `rangeInMemory` methods in the blockstate and blocktree classes and several others. WDYT?